[core]: support rename primary key columns and bucket key columns

zhongyujiang opened the pull request:
<!-- Please specify the module before the PR name: [core] ... or [flink] ... -->

### Purpose

<!-- Linking this pull request to the issue -->
Support renaming primary keys and bucket keys that are not partition keys.

<!-- What is the purpose of the change -->

### Tests

<!-- List UT and IT cases to verify this change -->
Added schema evolution tests and Spark e2e tests

### API and Format
<!-- Does this change affect API or storage format -->
Not affected
### Documentation

<!-- Does this change introduce a new feature -->

zhongyujiang commented:
cc @JingsongLi @FangYongs Could you please review this? Thanks

JingsongLi commented:
CC @tsreaper

zhongyujiang commented:
Gentle ping @JingsongLi @tsreaper, can you help review this when you have time? This can be useful to us, thanks!

JingsongLi reviewed (commented):
Hi @zhongyujiang , can you rebase master?

Can this support when table is empty?

zhongyujiang commented:
> Hi @zhongyujiang , can you rebase master?

Rebased. 
> Can this support when table is empty?

I guess this is a typo, what you mean is 'non-empty'? 
It appears that the bucket key and primary key are also projected based on the column name mapping, so renaming them does not affect the reading of old data. For this, I have added an end-to-end test `testRenamePrimaryKey `.

zhongyujiang commented:
> > Can this support when table is empty?
> 
> I guess this is a typo, what you mean is 'non-empty'?
> It appears that the bucket key and primary key are also projected based on the column name mapping, so renaming them does not affect the reading of old data. For this, I have added an end-to-end test testRenamePrimaryKey .


org.apache.paimon.schema.SchemaEvolutionUtil#createDataProjection

I have investigate more to confirm this. 
Before reading the data files, the Reader maps the primary key (and other value columns) to the actual columns in the data files. Therefore, I believe that renaming the primary key should not affect the reading of historical data, as long as the partition columns are not renamed. This is because we rely on them when calculating the paths. 
The relevant code can be found in org.apache.paimon.schema.SchemaEvolutionUtil#createDataProjection.

zhongyujiang commented on paimon-spark/paimon-spark-common/src/test/java/org/apache/paimon/spark/SparkSchemaEvolutionITCase.java:
```diff
@@ -276,7 +276,72 @@ public void testDropPrimaryKey() {
+
+        assertThat(actual).containsExactlyInAnyOrder("[1,aaa]", "[2,bbb]");
+
+        spark.sql("INSERT INTO test_rename_primary_key_table VALUES(1, 'AAA'), (2, 'BBB')");
```
This refers to the actual column projection used when reading data files during the debugging process.

This refers to data written before rename, the projection columns are `_KEY_a` and `a`:
![Screenshot 2024-08-04 at 21 51 16](https://github.com/user-attachments/assets/acfaaeff-667f-463e-9fa2-9818394a6fa9)

This refers to data written after rename, the projection columns are `_KEY_a_` and `a_`:
![Screenshot 2024-08-04 at 22 11 52](https://github.com/user-attachments/assets/8df1dace-703a-43ee-bc3a-838c191859e5)

zhongyujiang commented:
The CI failure seems caused by flaky tests, unreleated to this, I've filed a issue on that: #3908

JingsongLi commented:
> > Hi @zhongyujiang , can you rebase master?
> 
> Rebased.
> 
> > Can this support when table is empty?
> 
> I guess this is a typo, what you mean is 'non-empty'? It appears that the bucket key and primary key are also projected based on the column name mapping, so renaming them does not affect the reading of old data. For this, I have added an end-to-end test `testRenamePrimaryKey `.

`the bucket key and primary key are also projected based on the column name mapping`.

It is true, but this is very dangerous and may lead to many bugs.

zhongyujiang commented:
> It is true, but this is very dangerous and may lead to many bugs.

Hi @JingsongLi Could you share your specific concerns? From what I've observed, renaming a primary key doesn't seem to differ from renaming other keys; it appears to be safe.

I believe renaming the primary key is a necessary part of full schema evolution, and we have such a requirement ourselves. So if you have any concerns, I would be more than happy to investigate further and try to address them. Thank you!

JingsongLi commented:
re-open to trigger tests.

JingsongLi commented:
+1
